### ISSUE TITLE
Canonicalise to `type: integer` or `type: number`, never both

### ISSUE THREAD

**Issue by Zac-HD** [Zac-HD/hypothesis-jsonschema#43]
Dealing with schemas that might contain `"type": ["integer", "number"]` makes various tricks like discarding non-type-relevant keywords harder than it needs to be.

But if you stop and think about it, this is also a silly type to represent: either it's a `number` type (which may be of Python type `int`, I should check that in `from_schema`) and we can discard the `integer` part as adding no information at all, or it only allows integral values and it might as well be of `type: integer` to express that.

Notes to self:

- If `multipleOf` is an integer, we can convert `type: number` to `type: integer`
- To make this easy, and generally canonicalise things, we should round-trip schemas through `encode_canonical_json` to e.g. convert all integer-valued floats to integers.
- Check for type-related logic elsewhere which has to deal with multiple numeric types in a schema (ie for keyword relevance detection, maybe other things?) and replace it with comments explaining why it's no longer needed

**Comment by Zac-HD** [comment]
https://github.com/Zac-HD/hypothesis-jsonschema/compare/master...numeric-tricks implements all these suggestions, plus some incidentals.  Turns out they give Hypothesis enough insight into the structure of the data that our tests now reliably rediscover schemas which require #42 to translate, so I'll wait on that PR :smile:

**Comment by Zac-HD** [comment]
Aaand, done.
